### PR TITLE
Fix boost compilation and linking issues with recent clang

### PR DESCRIPTION
**Pull request by j-hellenberg**
Fix two boost issues, compilation and linkings issues with clang 18+, and more.

 * We used `add_compile_options(-DBOOST_ASIO_DISABLE_CONCEPTS)` in cmake which led to `/usr/include/boost/asio/co_composed.hpp:849:13: error: 'async_operation' has not been declared`. Removing it works just fine, was apparently fixed in boost 1.73 (https://github.com/boostorg/beast/issues/1884#issuecomment-635855251). We require 1.81+.
 * Ignore a recent warning of Clang 18/19
 * With cmake 3.30+ we no longer need `FindModule` for boost:
    ```
    Upstream Boost 1.70
   and above provide a ``BoostConfig.cmake`` package configuration file.
   ``find_package(Boost CONFIG)`` finds the upstream package directly,
   without the find module.
   ```
 * fix templating issues in the Postgres protocol handler with clang 18+
 * updated CI  macs to Sonoma and changed linking of `hyriseSystemTest` which made problems with macOS (`ld` works, but ld64.lld has issues for whatever reason); using LLVM 19 now
 * update two third-party modules to get rid of some cmake warnings

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -181,7 +181,9 @@ find_package(Tbb REQUIRED)
 find_package(Readline REQUIRED)
 find_package(Curses REQUIRED)
 find_package(SQLite3 REQUIRED)
-find_package(Boost REQUIRED COMPONENTS container date_time)
+# Old versions of CMake used the FindBoost module, which is deprecated because find_package can now find the upstream package directly.
```
Comment (Bouncner) [inline CMakeLists.txt:184]: What exactly does "upstream package" mean here?

Comment can be written on two lines.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -42,10 +42,10 @@ endif()
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_STANDARD 20)
 
-# Hotfix for boost bug (https://github.com/boostorg/asio/issues/312):
-add_compile_options(-DBOOST_ASIO_DISABLE_CONCEPTS)
 SET(BOOST_MIN_VERSION "1.81.0")
-FIND_PACKAGE(Boost ${BOOST_MIN_VERSION} REQUIRED)
+# Old versions of CMake used the FindBoost module, which is deprecated because find_package can now find the upstream package directly.
```
Comment (Bouncner) [inline src/CMakeLists.txt:46]: Not yours, but: lower case method call? And also limit the boost packages to "container date_time"?

**Review by Bouncner — COMMENTED** [review]
Two tiny comments.

Looks good to me but we have some issues on the ARM mac that need to be resolved first.

File: CMakeLists.txt (hunk 1 above)
Comment (j-hellenberg) [reply CMakeLists.txt:184]: See [here](https://cmake.org/cmake/help/latest/policy/CMP0167.html)

**Comment by Bouncner** [comment]
Looks like some ncurses issue that I don't want to fix as it just works with LLVM 19. I might pull some changes from the Ubuntu24.04 branch into this branch here to get it running with LLVM 19.

**Comment by Bouncner** [comment]
@j-hellenberg: can you please self-review? :)

File: CMakeLists.txt [hunk 3]
```diff
@@ -55,6 +55,18 @@ else()
     compiler_not_supported("You are using an unsupported compiler (${CMAKE_CXX_COMPILER_ID})! Compilation has only been tested with Clang (Linux + OS X) and GCC (Linux).")
 endif()
 
+if ("${CMAKE_CXX_COMPILER_ID}" MATCHES ".*Clang")
+     if (CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 18.0 AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 20.0)
+         # Clang 18/19 warn when switch statements do not have a default case even when the entire enum is covered (see
+         # https://github.com/llvm/llvm-project/blob/7d9634e527fe52bf20a9036be6e5771f8fc4de17/clang/test/Sema/switch-default.cpp#L19).
+         # As clang also (correctly) complains when not all enum cases are handled, we disable this warning.
```
Comment (dey4ss) [inline CMakeLists.txt:62]: which warning? :b

File: CMakeLists.txt [hunk 4]
```diff
@@ -55,6 +55,18 @@ else()
     compiler_not_supported("You are using an unsupported compiler (${CMAKE_CXX_COMPILER_ID})! Compilation has only been tested with Clang (Linux + OS X) and GCC (Linux).")
```
Comment (dey4ss) [inline CMakeLists.txt:55]:
```suggestion
    compiler_not_supported("You are using an unsupported compiler (${CMAKE_CXX_COMPILER_ID})! Compilation has only been tested with Clang (Linux + macOS) and GCC (Linux).")
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -181,7 +193,11 @@ find_package(Tbb REQUIRED)
 find_package(Readline REQUIRED)
 find_package(Curses REQUIRED)
 find_package(SQLite3 REQUIRED)
-find_package(Boost REQUIRED COMPONENTS container date_time)
+
+# Old versions of CMake used the FindBoost module, which is deprecated because find_package can now find
+# the upstream package directly. Therefore, disable it explicitly.
+SET(BOOST_MIN_VERSION "1.81.0")
```
Comment (dey4ss) [inline CMakeLists.txt:199]:
```suggestion
set(BOOST_MIN_VERSION "1.81.0")
```

**Review by dey4ss — APPROVED** [review]
Small things. Could you also provide a meaningful title (and a short description if necessary)?

File: CMakeLists.txt (hunk 3 above)
Comment (dey4ss) [reply CMakeLists.txt]: This was exactly what I meant.

**Comment by Bouncner** [comment]
@dey4ss : description okay?

@j-hellenberg : anything to add?

**Comment by j-hellenberg** [comment]
> > @j-hellenberg : anything to add?
> 
> Add him as contributor?

Sure, why not :)

LGTM, commits should be squashed before merging here, but I think you do that anyway for each PR

**Pull request merged**